app/src/main/java/org/wikipedia/readinglist/database/ReadingListDbHelper.java

```diff
@@ -240,12 +266,52 @@ void addPagesToList(@NonNull SQLiteDatabase db, @NonNull ReadingList list, @NonN
         return addedTitles;
     }
 
+    public List<String> movePagesToListIfNotExist(long sourceReadingListId, @NonNull ReadingList list, @NonNull List<PageTitle> titles) {
+        SQLiteDatabase db = getWritableDatabase();
+        db.beginTransaction();
+        List<String> movedTitles = new ArrayList<>();
+        try {
+            for (PageTitle title : titles) {
+                if (getPageByTitle(db, list.id(), title) != null) {
+                    continue;
+                }
+                movePageToList(db, sourceReadingListId, list, title);
+                movedTitles.add(title.getDisplayText());
+            }
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+        if (!movedTitles.isEmpty()) {
+            SavedPageSyncService.enqueue();
+            ReadingListSyncAdapter.manualSync();
+        }
+        return movedTitles;
+    }
+
     private void addPageToList(SQLiteDatabase db, @NonNull ReadingList list, @NonNull PageTitle title) {
         ReadingListPage protoPage = new ReadingListPage(title);
         insertPageInDb(db, list, protoPage);
         WikipediaApp.getInstance().getBus().post(new ArticleSavedOrDeletedEvent(true, protoPage));
     }
 
+    private void movePageToList(SQLiteDatabase db, long sourceReadingListId, @NonNull ReadingList list, @NonNull PageTitle title) {
+        ReadingListPage sourceReadingListPage = getPageByTitle(db, sourceReadingListId, title);
+        ReadingList sourceReadingList = getListWithoutContentsById(db, sourceReadingListId);
+        if (sourceReadingListPage != null && sourceReadingList != null) {
+            addPageToList(db, list, title);
+            markPageForDeletion(db, sourceReadingList, sourceReadingListPage);
+            WikipediaApp.getInstance().getBus().post(new ReadingListSyncEvent());
+        }
+    }
+
+    private void markPageForDeletion(SQLiteDatabase db, @NonNull ReadingList list, @NonNull ReadingListPage page) {
```

Comment:
Can we just use the existing `markPagesForDeletion` function?

Reply:
Done!